examples/sir_hmc.py

```diff
@@ -0,0 +1,134 @@
+
+
+
+"""
+First consider a simple SIR model.
+
+Note 1. Even in this simple model we would need to fix Binomial.log_prob() to
+    return -inf for values outside of its support.
```

Comment:
in NumPyro, when `validate_args` is enabled, we throw a warning (instead of error) if the value is out-of-support and mask the log_prob with `-inf` at those values. Should we also do so in Pyro?

cc @neerajprad

Reply:
That seems to be a better default at least for HMC. The error happens in the distributions `__init__` in pytorch, and we'll need to override that in Pyro to have the same behavior. We can also change the behavior under `validate_args=False` and replace any `NaN` or out of support values with `-inf`, but that will be expensive.

Reply:
I think for this PR a cleaner solution is to implement different distributions with appropriate support, so that validation logic can still be enabled:
```py
class ExtendedBinomial(Binomial):
    support = constraints.integer
    def log_prob(self, value):
        result = super().log_prob(value)
        return result.masked_fill((value < 0) | (self.total_count < value), -math.inf)
```
Maybe these can belong in `pyro.contrib.epidemiology` for now?

Reply:
If it is only a few distributions, this seems like the cleanest approach. If there is more demand for something like this, we can probably modify `_validate_sample` to return a boolean mask  and use that in `log_prob` as a more lax form of sample validation, but that will require more pervasive changes.

Reply:
https://github.com/pyro-ppl/pyro/pull/2414